### DISCUSSION TITLE
Layer 3 DSR support

### DISCUSSION THREAD

**Discussion by Slepwin** [loxilb-io/loxilb#216] [Ideas]
Hi,

Maybe this feature already in your roadmap, but anyway please add support for Layer 3 DSR mode.
It's very flexible and scalable solution in conjunction with BGP. All Linux distribution already can decapsulate IPIP tunnels just with enabled tunnel interface (without IP) and dummy interface (VIP address) + some arp and rp_filter tweaks.
Moreover if you add L3 DSR mode need do the same tunneling for endpoint probes (health-checks) to ensure data and health-checks traffic sharing the same network path, this prevent data traffic blackholing or faliled health-checks when networks issue happens.

**Comment by UltraInstinct14** [comment]
Thanks for all your review comments.  We will prioritize and work on these items asap.

**Comment by UltraInstinct14** [comment]
@Slepwin Support for the requested feature has been [added](https://github.com/loxilb-io/loxilb/issues/219). We are undergoing some further tests and will soon have a minor release for this. Also, a corresponding [scenario](https://github.com/loxilb-io/loxilb/blob/main/cicd/tcplbl3dsr/config.sh) has been added to our CICD pipeline for verification. Request to double check in the meantime.

**Reply by Slepwin** [reply]
Hi, team!

Many thanks for L3-DSR support. One more thing: as it mentioned in https://www.loxilb.io/post/loxilb-anycast-service-load-balancing-with-high-availability you should configure ip-ip tunnel for each endpoint and it's add complexity.
Example:
ip link add name ipip11 type ipip local 31.31.31.1 remote 31.31.31.253
$ip link set dev ipip11 up

But we have another option just enable Tunneling on endpoint without configuring point to point tunnels:
cat /etc/sysconfig/network-scripts/ifcfg-tunl0
DEVICE=tunl0
ONBOOT=yes

net.ipv4.conf.all.arp_ignore=1
net.ipv4.conf.all.arp_announce=2
net.ipv4.conf.all.rp_filter=0
net.ipv4.conf.tunl0.rp_filter=0

cat /etc/sysconfig/network-scripts/ifcfg-lo:qq
DEVICE=lo:qq
IPADDR=x.x.x.x
NETMASK=255.255.255.255
ARP=no
ONBOOT=yes
NAME=loopback

In that configuration every encapsulated packet would be decapsulated on endpoint host (tested with IPVS as LB and Ubuntu/RHEL based OS).
So my requests: 
1. add auto encapsilation on loxilb (like IPVS kernel module in TUN mode) without the need to configure tunnels in Linux.
2. add auto encapsilation on loxilb for healthchecks too to ensure VIP (on dummy/loopback interfaces), rp_filter and arp sysctl settings configured properly.

Additional notes: With autotunneling IPVS used VIP address as dst IP in inner header and endpoint IP as dst IP in outer header.

**Reply by iluffk** [reply]
@TrekkieCoder we use the same (with ipvs), it would be good to implement that approach with ebpf
